lib/bamboo/adapters/sendgrid_adapter.ex

```diff
@@ -0,0 +1,160 @@
+defmodule Bamboo.SendgridAdapter do
+  @moduledoc """
+  Sends email using SendGrid's JSON API.
+
+  Use this adapter to send emails through SendGrid's API. Requires that an API
+  key is set in the config.
+
+  ## Example config
+
+      # In config/config.exs, or config.prod.exs, etc.
+      config :my_app, MyApp.Mailer,
+        adapter: Bamboo.SendgridAdapter,
+        api_key: "my_api_key"
+
+      # Define a Mailer. Maybe in lib/my_app/mailer.ex
+      defmodule MyApp.Mailer do
+        use Bamboo.Mailer, otp_app: :my_app
+      end
+  """
+
+  @default_base_uri "https://api.sendgrid.com/api"
+  @send_message_path "/mail.send.json"
+  @behaviour Bamboo.Adapter
+
+  alias Bamboo.Email
+
+  defmodule ApiError do
+    defexception [:message]
+
+    def exception(%{params: params, response: response}) do
+      filtered_params = params |> Plug.Conn.Query.decode |> Map.put("key", "[FILTERED]")
+
+      message = """
+      There was a problem sending the email through the SendGrid API.
+
+      Here is the response:
+
+      #{inspect response, limit: :infinity}
+
+
+      Here are the params we sent:
+
+      #{inspect filtered_params, limit: :infinity}
+      """
+      %ApiError{message: message}
+    end
+  end
+
+  def deliver(email, config) do
+    api_key = get_key(config)
+    body = email |> to_sendgrid_body |> Plug.Conn.Query.encode
+
+    case HTTPoison.post!(base_uri <> @send_message_path, body, headers(api_key)) do
+      %{status_code: status} = response when status > 299 ->
+        raise(ApiError, %{params: body, response: response})
+      response -> response
+    end
+  end
+
+  @doc false
+  def handle_config(config) do
+    if config[:api_key] in [nil, ""] do
+      raise_api_key_error(config)
+    else
+      config
+    end
+  end
+
+  defp get_key(config) do
+    case Map.get(config, :api_key) do
+      nil -> raise_api_key_error(config)
+      key -> key
+    end
+  end
+
+  defp raise_api_key_error(config) do
+    raise ArgumentError, """
+    There was no API key set for the SendGrid adapter.
+
+    * Here are the config options that were passed in:
+
+    #{inspect config}
+    """
+  end
+
+  defp headers(api_key) do
+    [
+      {"Content-Type", "application/x-www-form-urlencoded"},
+      {"User-Agent", "bamboo/#{Bamboo.version}"},
```

Comment:
I think it might be better to remove the stuff from `mix.exs` and instead set a module attribute in this module `@sendgrid_adapter_version`. I think that's more clear. What do you think?